ansible_galaxy/rest_api.py

```diff
@@ -247,3 +252,44 @@ def fetch_content_related(self, related_url):
             done = (data.get('next_link', None) is None)
 
         return results
+
+    @g_connect
+    def fetch_namespace(self, namespace):
+        namespace = urlquote(namespace)
+        url = '%s/namespaces/?name=%s' % (self.baseurl, namespace)
+        data = self.__call_galaxy(url, http_method='GET')
+        if data["results"]:
+            return data["results"][0]
+        return {}
+
+    @g_connect
+    def publish_file(self, namespace_id, data, archive_path):
+        form = MultiPartForm()
+        for key in data:
+            form.add_field(key, data[key])
+
+        form.add_file('file', os.path.basename(archive_path),
+                      fileHandle=codecs.open(archive_path, "rb"))
+
+        url = '%s/namespaces/%s/collection' % (self.baseurl, namespace_id)
+        _, netloc, url, _, _, _ = urlparse(url)
+        try:
+            form_buffer = form.get_binary().getvalue()
+            http = http_client.HTTPConnection(netloc)
```

Comment:
What is the reason for using http client instead of ansible_galaxy.flat_rest_api.urls.open_url? 
In theory that should supports POST requests (but can't say I've actually tried to use it).

Reply:
I tried calling `self.__call_galaxy()` in `rest_api.py`. However, urllib2 will only transmit string data. Even if the payload is assembled as a string, the contents of the archive file is binary, and urllib still pukes.

Passing it binary data, causes `httplib.py` to throw `UnicodeDecodeError: 'ascii' codec can't decode byte 0x8b in position 1107: ordinal not in range(128)`

As suggested by [this blog post](https://blog.thesparktree.com/the-unfortunately-long-story-dealing-with), I bypassed urllib2, and called httplib directly.